python-oop/static-methods-and-class-methods.md

```diff
@@ -0,0 +1,268 @@
+
+Even if we call a class method from an instance, we will not have access to the instance it`self`.
+
+Observe this code, which shows that `cls` is still the value of the class, no matter what invokes the class method.
+
+```python
+class ExampleClass:
+
+    @classmethod
+    def example_method(cls):
+        print("I'm inside the class method, example_method!")
+        print("In a class method, cls will be the class itself", cls)
+```
+
+```python
+example_instance = ExampleClass()
+example_instance.example_method()
+```
+
+Output:
+
+```
+I'm inside the class method, example_method!
+In a class method, cls will be the class itself <class '__main__.ExampleClass'>
+```
+
+We should observe from the output that:
+
+- The code successfully ran the class method, even though it was invoked with `example_instance.example_method()`
+- When we printed `cls`, it continued to be `<class '__main__.ExampleClass'>`
+
+### Using `cls` to Access Class Variables
+
+As mentioned above, class methods can access any class variables that are defined, using the `cls` variable:
+
+```python
+class ExampleClass:
+    example_class_var = "This is an example class variable!"
+
+    @classmethod
+    def example_method(cls):
+        print("I'm inside the class method, example_method!")
+        print("I can access class variables using the cls parameter:", cls.example_class_var)
+```
+
+```
+ExampleClass.example_method()
+```
+
+This code produces the console output:
+
+```
+I'm inside the class method, example_method!
+I can access class variables using the cls parameter: This is an example class variable!
+```
+
+Of course, class variables can be accessed without the `cls` parameter, as long as there's access to the class itself. The following version of `ExampleClass` produces the same output as the previous version:
+
+```python
+class ExampleClass:
+    example_class_var = "This is an example class variable!"
+
+    @classmethod
+    def example_method(cls):
+        print("I'm inside the class method, example_method!")
+        print("I can access class variables using the cls parameter:", ExampleClass.example_class_var)
+```
+
+The difference is that `cls` will _always_ represent the current class itself; specifying `ExampleClass` could have consequences if class definitions change.
+
+## Class Methods: Application
+
+Class methods are special because they have access to class variables, and class state.
+
+Class methods are great for:
+
+- modifying class state that applies across all instances of the class
+- doing operations that aren't associated with an instance
+- doing operations that rely on the class itself, such as making instances of that class
+
+## Check for Understanding
+
+<!-- Use the static method decorator to create a static method that... -->
+<!-- Use the class method decorator to create a static method that... -->
```

Comment:
```suggestion
<!-- prettier-ignore-start -->
### !challenge

* type: checkbox
* id: 6d4dab23-bc9d-45fa-a4d8-07689ecf84c7
* title: Static Methods and Class Methods I 

##### !question
Select the options that complete this sentence:

> A static method ..
##### !end-question

##### !options
* is bound to a class
* does not have access to instance variables (attributes) or class variables. 
* can access or modify class state 
* requires `cls` as first argument
##### !end-options

##### !answer
* is bound to a class
* does not have access to instance variables (attributes) or class variables.
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge

* type: checkbox
* id: 6d4dab23-bc9d-45fa-a4d8-07689ecf84c7
* title: Static Methods and Class Methods II

##### !question
Select the options that complete this sentence:

> A class method ..
##### !end-question

##### !options
* is bound to a class
* has access to instance variables (attributes) or class variables. 
* can access or modify class state 
* requires `cls` as first argument
##### !end-options

##### !answer
* is bound to a class
* has access to instance variables (attributes) or class variables. 
* can access or modify class state 
* requires `cls` as first argument
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: 3c69fae9-f63c-42f4-915a-9fca215bc374
* title: Static Methods and Class Methods III

##### !question
Select the decorator that would call the `subtractNumbers` method and print `5` to the console.

``` Python
class Calculator:
    def subtractNumbers(x, y):
        return x - y
print(Calculator.subtractNumbers(10,5))
```
##### !end-question

##### !options
* @staticmethod
* @classmethod
##### !end-options

##### !answer
* @staticmethod
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: checkbox
* id: 1d5979f7-0617-4a16-88f7-2da87c6c5a75
* title: Static Methods and Class Methods IV

##### !question
Select the option that will print the following output:

```
Your total will be $20.
Ask about our membership program to receive a 10% discount today!
```
##### !end-question

##### !options
* 
``` Python
class CalculateTotal:
    summer_discount = 10
    winter_discount = 15
    @classmethod
    def promote_membership(cls):
        print(f"Ask about our membership program to receive a {cls.summer_discount}'%' discount today!")

def display_total(amount):
    print(f"You total will be ${amount}")
    print(CalculateTotal.promote_membership())

display_total(20)
```
* 
``` Python
class CalculateTotal:
    summer_discount = 10
    winter_discount = 15
    
    @classmethod
    def promote_membership(cls):
        return "Ask about our membership program to receive a " + str(CalculateTotal.summer_discount) + "% discount today!"

def display_total(amount):
    print(f"Your total will be ${amount}")
    print(CalculateTotal.promote_membership())

display_total(20)
```

##### !end-options

##### !answer
* 
``` Python
class CalculateTotal:
    summer_discount = 10
    winter_discount = 15
    @classmethod
    def promote_membership(cls):
        print(f"Ask about our membership program to receive a {cls.summer_discount}'%' discount today!")

def display_total(amount):
    print(f"You total will be ${amount}")
    print(CalculateTotal.promote_membership())

display_total(20)
```
##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->
```

Reply:
Problem 2 (Static Methods and Class Methods II):
- includes "has access to instance variables (attributes) or class variables." but it does _not_ have access to instance variables.  This should just be "has access to class variables, but not instance variables."

Problem 4 (Static Methods and Class Methods IV):
- has a typo in the answer output "You"
- has weird stuff going on with an extra print in the classmethod
- the "wrong" answer prints the same correct answer string